### PR TITLE
libsystem: Add MCSFileGetType().

### PR DESCRIPTION
**Pull request by peter-b**
Basic ability to distinguish between files and directories, given a
path.

### CODE REVIEW

**Comment by livecodestephen** [comment]
I see you throw an exception for unsupported file types.  Would it be better to include an other type in the enum?  I just thinking of the use case where you've got a path from the user and want to know what type it is.  As it stands you'd have to put this in a try block, which increases the complexity of the code, when really this is just another case.

**Comment by peter-b** [comment]
@livecodestephen: At the moment the idea was just to implement the absolute minimum functionality required for getting the LC8 test suite working again on Linux, OS X and Windows (and nothing more).  Really, there should be `kMCSFileType` enum entries for every filesystem node type that can be encountered across all platforms.  Having the exception there for the time being is useful as a reminder that this thing actually needs to be extended/replaced at some point.

TL;DR This feature is not finished.

**Comment by runrevmark** [comment]
@peter-b: Wouldn't an 'unknown' return value make more sense... After all at any given point in time any one OS might decide to introduce a new file type which any existing app would not understand. Therefore, by adding an 'unknown' return value type, it 'forces' the app using the API to handle 'unknown', meaning ti wouldn't break if such a case were to occur.

_mini-bikeshed_

**Comment by peter-b** [comment]
@runrevmark Bikeshed indeed, but it's worth getting right.  So, my question is this: how do you distinguish between "this is a type of file that libsystem doesn't know about" and "this is a type of file that I don't handle want to handle"?  I think that the "if this is file is of a type that `MCSFileGetType()` doesn't know about" is an odd idiom...

In my use case -- I'd actually rather get an exception (which is why I wrote it that way).  libsystem currently generally tries to bail out early for things it can't support, rather than muddling along.

Finally, just because your code handles "unknown", that doesn't really tell you anything useful (because if we add a new entry to the enumeration, your code just broke).  If you `switch` over `MCSFileType` you _must_ include a `default` clause anyway...

**Comment by peter-b** [comment]
TBH I need to get this merged so I can get the LC8 testsuite working again, so if a review's going to be held hostage over a `MCSFileTypeUnknown` constant, I guess I'll have to go and add it. :stuck_out_tongue:

**Comment by runrevmark** [comment]
I'm not going to hold anything hostage over it - I was just proposing a slightly different way of thinking about things for the future.

The current implementation is obviously fine for its current intended purpose - where we _know_ what we are looking for and if it is not the thing we are expecting then the structure of our unittest folders are wrong - thus it should be an error.

I was suggesting that throwing an exception on a type of file which libsystem does not know about is probably not the correct approach in the future. It prevents you from being able to write fault-tolerant code which uses it. e.g.

I write a neat little tool in LiveCode which iterates over the filesystem to present useful info. I take that tool on my system and it works really well. I take it over to Bob's machine. Unfortunately Bob likes hacking around with Linux and playing around with ideas - his system has a file type which mine doesn't, nor one which libsystem understands. My tool breaks. I cannot fix nor make it work on his system. Why? Because I made the assumption that something that libsystem does not know about cannot exist and therefore looking at it must be an error. Whereas, in fact, the flaw is that libsystem didn't provide me with the choice about what to do should it encounter something which it doesn't know about.

In terms of the switch/default - that's not really relevant. My suggestion was to add an explicit 'I don't know' to the enum which apps can then handle. A default clause would do the same thing - but there still needs to be an explicit value which libsystem can return to say it doesn't know, even if apps always handle that in default. (At them the default clause would never trigger for a fully switched MCSFileType).

**Comment by peter-b** [comment]
@runrevmark @livecodestephen: Pull request updated. :wink:

**Comment by runrevmark** [comment]
(I should perhaps point out that - of course - the chance of finding a Linux system with a new FS type is very very low... It was more an API design question - the better APIs I've come across which enumerate / classify things always have an 'Unknown' in any enum for a type of thing. It makes writing code that uses them easier as you can then choose whether to handle Unknown as an error, or to handle it in some special way in the same way as other filetypes. As Stephen pointed out, if its an error then you need a sideline mechanism to process it if you want to handle 'unknown' as part of everything else).

**Comment by runrevmark** [comment]
@livecode-vulcan review ok 1c6c92f

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 1c6c92f0dd3b7f1c37acbef25d968febf6b93452

**Comment by livecode-vulcan** [comment]
😞 test failure 1c6c92f0dd3b7f1c37acbef25d968febf6b93452
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/36)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/36)
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/36)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/36)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/38)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/36)

**Comment by peter-b** [comment]
@runrevmark: Pull request updated to fix build failure on win-x86.

**Comment by runrevmark** [comment]
@livecode-vulcan review ok 5fc7c3d

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 5fc7c3ddf1127623a60a783bfd5f28e8e2115a0a

**Comment by livecode-vulcan** [comment]
😎 test success 5fc7c3ddf1127623a60a783bfd5f28e8e2115a0a
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/37)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/37)
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/37)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/37)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/39)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/37)
